### PR TITLE
Fix toplevel transpile

### PR DESCRIPTION
**Pull request by aethos**
makes testing slow but makes the toplevel do the right stuff

now toplevel will compile your file for you and generate an executable
eg: `./oscar -c divide.oscar -o divide_exec`
will generate the cpp file `divide.cpp` and executable `divide_exec`

also added an error for scanners

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -16,7 +16,9 @@ OCAMLC=ocamlfind
 
 BUILDDIR=include
 
-oscar :
+OBJS = ast.cmx codegen.cmx parser.cmx scanner.cmx semant.cmx oscar.cmx
```
Comment (Howon) [inline Makefile]: why

File: src/oscar.ml [hunk 2]
```diff
@@ -6,47 +6,56 @@ open Sast
...
+    (fprintf stderr "Parser: Syntax error at line %d, char %d - %d: \"%s\" \n"
     s.pos_lnum (s.pos_cnum - s.pos_bol)
     (f.pos_cnum - f.pos_bol) (Lexing.lexeme lexbuf))
 
 let _ =
   let arg_len = Array.length Sys.argv in
   let (action, oscar) =
-    (if (arg_len = 1 || arg_len > 3) then
-      let _ = print_endline("Usage: ./oscar [-l|-c|-s] *.oscar") in
+    (if (arg_len < 3 || arg_len = 4 || arg_len > 5) then
```
Comment (Howon) [inline src/oscar.ml:34]: why

File: src/transpile.ml [hunk 3]
```diff
@@ -45,7 +45,7 @@ and c_texpr tex =
     (match se with
         SInt_Lit _ | SDouble_Lit _ | SChar_Lit _
       | SBool_Lit _ | SId _ | SNoexpr -> str_texpr tex
-      | SUnit_Lit _                -> "void"
+      | SUnit_Lit _                -> "unit"
```
Comment (Howon) [inline src/transpile.ml]: why

File: src/transpile.ml [hunk 4]
```diff
@@ -93,9 +93,9 @@ and c_formals fs = String.concat "," (List.map c_formal fs)
 
 and c_lambda sfl =
   let { sf_formals = sformals; sf_return_t = srt; sf_body = sbody } = sfl in
-    "[&](" ^ c_formals sformals ^ ")" ^ (match srt with
-        Unit_t -> ""
-      | _ -> "->" ^ c_type srt) ^ c_sstmt sbody;
+    "[]()" ^ (match srt with
```
Comment (Howon) [inline src/transpile.ml]: why

File: src/transpile.ml [hunk 5]
```diff
@@ -93,9 +93,9 @@ and c_formals fs = String.concat "," (List.map c_formal fs)
...
+          Unit_t -> ""
```
Comment (Howon) [inline src/transpile.ml]: why

File: src/transpile.ml [hunk 6]
```diff
@@ -93,9 +93,9 @@ and c_formals fs = String.concat "," (List.map c_formal fs)
...
+        | _ -> "->" ^ c_type srt) ^ c_sstmt sbody;
```
Comment (Howon) [inline src/transpile.ml]: why

**Comment by Howon** [comment]
Why are you overriding @jibbenHillen 's lambda functions purposefully?

File: test/oscar/compiler/mapPut.oscar [hunk 7]
```diff
@@ -3,4 +3,4 @@ def main() => unit = {
   Println(Put(2, -1.3, tMap) == map<int, double>[0 -> 1.1, 4 -> 5.3, -3 -> 5.3, 2 -> -1.3]);
   Println(tMap == map<int, double>[0 -> 1.1, 4 -> 5.3, -3 -> 5.3]);
   Println(Put(0, 1.1, tMap) == map<int, double>[0 -> 1.1, 4 -> 5.3, -3 -> 5.3]);
-}
+}
```
Comment (Howon) [inline test/oscar/compiler/mapPut.oscar]: p sure this is gnna break our parser

File: src/oscar.ml [hunk 8]
```diff
@@ -60,14 +69,35 @@ let _ =
...
+                      | None -> raise (Failure ("Filename should be .oscar")))
+                  in
+                  String.sub oscar 0 rdot
+              in
+              let cpp_file = file_stub ^ ".cpp" in
+              match action with
+                  Ast      -> ()
+                | Sast     -> ()
+                | Llvm_gen -> 
+                    let c_op = "-Wall -pedantic -fsanitize=address -std=c++1y -O2 -S -emit-llvm" in
```
Comment (Howon) [inline src/oscar.ml:88]: lets put this at the top

**Comment by Howon** [comment]
some of the tests you opened might have parsing errors

File: src/oscar.ml (hunk 2 above)
Comment (aethos) [reply src/oscar.ml:34]: Warning 3: deprecated: Core.Std.!=
[since 2014-10] Use [phys_equal]

**Review by Howon — APPROVED** [review]

**Pull request merged**